### PR TITLE
`jax.jit` works with `qml.probs` and finite shots

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
`qml.math.unique` cannot be jitted with JAX because the output shape of `unique` depends on the input values.

**Description of the Change:**
Use an input agnostic way to update the probs.

**Benefits:**
`qml.probs` can be jitted (without Python callbacks).

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by vincentmr** [comment]
[sc-62385]

File: tests/measurements/test_probs.py [hunk 1]
```diff
@@ -226,6 +228,36 @@ def circuit():
...
+            qml.PhaseShift(x, wires=1)
+            qml.CNOT(wires=[0, 1])
+            if isinstance(obs, Sequence):
+                return qml.probs(wires=obs)
+            return qml.probs(op=obs)
+
+        # expected probability, using [00, 01, 10, 11]
+        # ordering, is [0.5, 0.5, 0, 0]
+
+        assert "pure_callback" not in str(jax.make_jaxpr(circuit)(params))
```
Comment (vincentmr) [inline tests/measurements/test_probs.py:255]: `pure_callback` should be absent from `jaxpr` to make sure we are not using Python callbacks.

File: pennylane/measurements/probs.py [hunk 2]
```diff
@@ -259,23 +259,21 @@ def _count_samples(indices, batch_size, dim):
...
-            for b, idx in enumerate(indices):
-                basis_states, counts = qml.math.unique(idx, return_counts=True)
-                prob[basis_states, b] = counts / bin_size
+        interface = qml.math.get_deep_interface(indices)
 
-            return prob
+        def _count_samples_core(indices, dim, interface):
+            return qml.math.array(
+                [[qml.math.sum(idx == p) for p in range(dim)] for idx in indices], like=interface
+            )
```
Comment (dwierichs) [inline pennylane/measurements/probs.py:267]: Could this "dense" iteration over the `dim` range become a bottleneck? :thinking: In the previous version of the code, `dim` could be very large without affecting the performance significantly, if the `indices` are sparsely populated, I think?

File: pennylane/measurements/probs.py [hunk 3]
```diff
@@ -259,23 +259,21 @@ def _count_samples(indices, batch_size, dim):
...
-            # count the basis state occurrences, and construct the probability vector for each bin
-            for b, idx in enumerate(indices):
-                basis_states, counts = qml.math.unique(idx, return_counts=True)
-                prob[basis_states, b] = counts / bin_size
+        interface = qml.math.get_deep_interface(indices)
 
-            return prob
+        def _count_samples_core(indices, dim, interface):
+            return qml.math.array(
+                [[qml.math.sum(idx == p) for p in range(dim)] for idx in indices], like=interface
```
Comment (dwierichs) [inline pennylane/measurements/probs.py:266]: If we `swapaxes` on any usage of `_count_samples_core` anyways, couldn't we do this and skip the swapping?
```suggestion
                [[qml.math.sum(idx == p) for idx in indices] for p in range(dim)], like=interface
```

File: tests/devices/default_qubit/test_default_qubit_native_mcm.py [hunk 4]
```diff
@@ -695,11 +696,12 @@ def func(x, y, z):
...
 
     measures = [
-        # qml.probs,
+        qml.probs,
+        qml.probs,
         qml.sample,
         qml.sample,
         qml.expval,
-        # qml.probs,
+        qml.probs,
```
Comment (dwierichs) [inline tests/devices/default_qubit/test_default_qubit_native_mcm.py:704]: :100:

File: doc/releases/changelog-dev.md [hunk 5]
```diff
@@ -89,6 +89,9 @@
 
 <h3>Bug fixes 🐛</h3>
 
+* Finite shot circuits with a `qml.probs` measurement with a `wires` or `op` argument can now be compiled with `jax.jit`.
```
Comment (dwierichs) [inline doc/releases/changelog-dev.md:92]:
```suggestion
* Finite shot circuits with a `qml.probs` measurement, both with a `wires` or `op` argument, can now be compiled with `jax.jit`.
```

**Review by dwierichs — COMMENTED** [review]
Nice job @vincentmr , thanks!

Just one question regarding performance and one small simplification suggestion.

File: pennylane/measurements/probs.py (hunk 3 above)
Comment (vincentmr) [reply pennylane/measurements/probs.py]: I think I went with that to loop over the contiguous index, but it's probably pointless for a Python list.

File: pennylane/measurements/probs.py (hunk 2 above)
Comment (vincentmr) [reply pennylane/measurements/probs.py]: Good point, let's keep the old implementation when not tracing.

File: pennylane/measurements/probs.py (hunk 3 above)
Comment (dwierichs) [reply pennylane/measurements/probs.py]: Ah, good thinking! I would _think_ that for lists it indeed doesn't matter :thinking:

File: pennylane/measurements/probs.py [hunk 6]
```diff
@@ -259,23 +259,28 @@ def _count_samples(indices, batch_size, dim):
...
+                    like=interface,
+                )
+            probabilities = qml.math.zeros((dim, num_bins), dtype="float64")
             for b, idx in enumerate(indices):
                 basis_states, counts = qml.math.unique(idx, return_counts=True)
-                prob[basis_states, b] = counts / bin_size
-
-            return prob
+                probabilities[basis_states, b] = counts
+            return probabilities
```
Comment (dwierichs) [inline pennylane/measurements/probs.py:274]: As we know `indices` at the time we define this function, we could also define the function differently, conditioned on whether `qml.is_abstract(indices)`, rather than having one function that decides between the two cases.
This is a rather subjective suggestion, so feel free not to go with it!

**Review by dwierichs — APPROVED** [review]
Thanks @vincentmr!
Added a small comment regarding an optional, subjective change.

File: pennylane/measurements/probs.py (hunk 6 above)
Comment (vincentmr) [reply pennylane/measurements/probs.py:274]: Do you mean something like
```
        if qml.math.is_abstract(indices):
            def _count_samples_core(indices, dim, interface):
                return qml.math.array(
                    [[qml.math.sum(idx == p) for idx in indices] for p in range(dim)],
                    like=interface,
                )
        else:
            def _count_samples_core(indices, dim, interface):
                probabilities = qml.math.zeros((dim, num_bins), dtype="float64")
                for b, idx in enumerate(indices):
                    basis_states, counts = qml.math.unique(idx, return_counts=True)
                    probabilities[basis_states, b] = counts
                return probabilities
```
?

Comment (dwierichs) [reply pennylane/measurements/probs.py:274]: Yep, exactly :)

**Review by dwierichs — APPROVED** [review]
Thanks @vincentmr !

File: doc/releases/changelog-dev.md [hunk 7]
```diff
@@ -89,6 +89,9 @@
 
 <h3>Bug fixes 🐛</h3>
 
+* Finite shot circuits with a `qml.probs` measurement, both with a `wires` or `op` argument, can now be compiled with `jax.jit`.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:92]:
```suggestion
* Finite shot circuits with a `qml.probs` measurement, both with a `wires` or `op` argument, can now be fully compiled with `jax.jit` without use of inadvertent pure callbacks.
```


The current phrasing kinda makes it sound like probs+jax.jit straight up didn't work.  But it did; it was just less than inefficient.

Comment (vincentmr) [reply doc/releases/changelog-dev.md:92]: It straight up didn't work because of `qml.math.unique`. It now works and is free of callbacks.

**Review by albi3ro — APPROVED** [review]
